### PR TITLE
Update postgres query interface

### PR DESCRIPTION
**Pull request by turbofoxwave**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)
I have 1 failing testing on my system with the current master branch before adding my changes 
Failing test:
[POSTGRES] DataTypes should parse an empty GEOMETRY field:
     SequelizeDatabaseError: Too few ordinates in GeoJSON
All others are passing.

- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
no

- [x] Have you added new tests to prevent regressions? yes
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?  na
- [x] Have you added an entry under `Future` in the changelog? yes

### Description of change
Changes are around the postgres query generator to address some issues I found when attempting to use dropFunction and createFunction. The root cause of the issue was due to the function expandFunctionParamList. Along the way I also found an issue with dropFunction in which the generated sql lacked a space after the keyword 'FUNCTION' ex: "FUNCTION myfunc" vs "FUNCTIONmyfunc".  I've also added tests for createFunction and dropFunction and I've stubbed out some tests for renameFunction.

### CODE REVIEW

File: lib/dialects/postgres/query-generator.js [hunk 1]
```diff
@@ -524,6 +524,10 @@ const QueryGenerator = {
   },
 
   createFunction(functionName, params, returnType, language, body, options) {
+    if(!returnType) throw new Error("requires returnType");
```
Comment (sushantdhiman) [inline lib/dialects/postgres/query-generator.js]: Spaces around if condition something like ``if (!returnType) {``, I feel you can combine all three `returnType` , `language` and `body` checks and throw a generic error message like "createFunction missing some parameters, did you forget to pass returnType, language or body."

File: lib/dialects/postgres/query-generator.js [hunk 2]
```diff
@@ -564,24 +569,32 @@ const QueryGenerator = {
...
+    const paramList = [];
+    Utils._.each(params, curParam => {
       const paramDef = [];
       if (Utils._.has(curParam, 'type')) {
         if (Utils._.has(curParam, 'direction')) { paramDef.push(curParam.direction); }
         if (Utils._.has(curParam, 'name')) { paramDef.push(curParam.name); }
         paramDef.push(curParam.type);
       } else {
-        throw new Error('createFunction called with a parameter with no type');
+        throw new Error('expandFunctionParamList: parameter missing type');
```
Comment (sushantdhiman) [inline lib/dialects/postgres/query-generator.js]: We don't need to change this error message, `expandFunctionParamList` is a private api that is used by `createFunction`.

File: package.json [hunk 3]
```diff
@@ -113,6 +113,7 @@
     "test-integration-mysql": "DIALECT=mysql npm run test-integration",
     "test-integration-postgres": "DIALECT=postgres npm run test-integration",
     "test-integration-postgres-native": "DIALECT=postgres-native npm run test-integration",
+    "test-integration-query-interface-postgres": "DIALECT=postgres ./node_modules/.bin/mocha --globals setImmediate,clearImmediate --ui tdd --check-leaks --colors -t 15000 --reporter spec 'test/integration/dialects/postgres/query-interface.test.js'",
```
Comment (sushantdhiman) [inline package.json]: We don't need to add separate test command for `query interface`, we have integration test commands that will run those tests.

File: lib/dialects/postgres/query-generator.js [hunk 4]
```diff
@@ -564,24 +569,32 @@ const QueryGenerator = {
...
       }
-      return paramDef.join(' ');
+
+      var joined = paramDef.join(' ');
+      return joined ?  paramList.push(joined): null;
```
Comment (sushantdhiman) [inline lib/dialects/postgres/query-generator.js]: if `joined` exists only push to `paramList`, no `return` required now

File: test/integration/dialects/postgres/query-interface.test.js [hunk 5]
```diff
@@ -15,6 +15,190 @@ if (dialect.match(/^postgres/)) {
       this.queryInterface = this.sequelize.getQueryInterface();
     });
 
+
+    describe('createFunction', function(){
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: spaces in function expression, `function () {` , This applies to all other functions too

File: test/integration/dialects/postgres/query-interface.test.js [hunk 6]
```diff
@@ -15,6 +15,190 @@ if (dialect.match(/^postgres/)) {
...
+
+      beforeEach(function(){
+        var self = this;
+        return self.queryInterface.dropFunction('create_job',[{type:'varchar',name:'test'}]).catch(function(err){
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: dont throw error in `beforeEach` , extract this test as separate `it` spec , Also use a better way to [check for errors](http://chaijs.com/api/bdd/#method_throw)

File: test/integration/dialects/postgres/query-interface.test.js [hunk 7]
```diff
@@ -15,6 +15,190 @@ if (dialect.match(/^postgres/)) {
...
+      });
+
+      it('creates a stored procedure', function(){
+
+        var self = this;
+        var body = 'return test;';
+        var options = {};
+        return self.queryInterface.createFunction('create_job', [{type:'varchar',name:'test'}], 'varchar', 'plpgsql', body, options)
+        .then(function(){
+          return self.sequelize.query('select create_job(\'test\');', { type: self.sequelize.QueryTypes.SELECT })
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: pyramid of doom :0

File: test/integration/dialects/postgres/query-interface.test.js [hunk 8]
```diff
@@ -15,6 +15,190 @@ if (dialect.match(/^postgres/)) {
...
+          });
+        });
+      });
+
+      it('requires parameters array', function(){
+        var self = this;
+        var body = 'return 1;';
+        var options = {};
+        try{
+          return self.queryInterface.createFunction('create_job',null, 'integer', 'plpgsql', body, options);
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: can be refactored to use the `.throws` helper from chaijs

File: lib/dialects/postgres/query-generator.js (hunk 2 above)
Comment (turbofoxwave) [inline lib/dialects/postgres/query-generator.js]: expandFunctionParamList appears to be usedby 4 different functions in query-generator.
createTrigger, createFunction, dropFunction, renameFunction.
What do you think about reducing it to "parameter missing type" ?

File: package.json (hunk 3 above)
Comment (turbofoxwave) [inline package.json]: understood, I'll remove this asap when I get home this evening.

File: lib/dialects/postgres/query-generator.js (hunk 2 above)
Comment (sushantdhiman) [inline lib/dialects/postgres/query-generator.js]: Use this message "function or trigger used with a parameter without any type"

File: lib/dialects/postgres/query-generator.js [hunk 9]
```diff
@@ -564,24 +566,32 @@ const QueryGenerator = {
...
         if (Utils._.has(curParam, 'name')) { paramDef.push(curParam.name); }
         paramDef.push(curParam.type);
       } else {
-        throw new Error('createFunction called with a parameter with no type');
+        throw new Error('expandFunctionParamList: parameter missing type');
       }
-      return paramDef.join(' ');
+
+      var joined = paramDef.join(' ');
+      paramList.push(joined);
```
Comment (sushantdhiman) [inline lib/dialects/postgres/query-generator.js]: `joined` can be empty, Use `if (joined) {` check

File: test/integration/dialects/postgres/query-interface.test.js [hunk 10]
```diff
@@ -15,6 +17,182 @@ if (dialect.match(/^postgres/)) {
...
+        });
+      });
+
+      it('treats options as optional', function () {
+        var self = this;
+        var body = 'return test;';
+
+        //setup - ensure we don't have a create_job when we start
+        return new Promise.method( function (){
+          return self.queryInterface.dropFunction('create_job',[{type:'varchar',name:'test'}])
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: setup and cleanup both should go in `beforeEach` and `afterEach`. But you don't need to throw any errors inside `beforeEach` or `afterEach`

File: test/integration/dialects/postgres/query-interface.test.js [hunk 11]
```diff
@@ -15,6 +17,182 @@ if (dialect.match(/^postgres/)) {
...
+            if (res[0].create_job !== 'test') throw new Error('expected output: test');
+          });
+        })
+        //cleanup
+        .then( function () {
+          return self.queryInterface.dropFunction('create_job',[{type:'varchar',name:'test'}]);
+        });
+      });
+
+      it('requires function name', function () {
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: all these require parameter test can be combined together with having mutiple `expect` to check for each case. Currently all these have exactly same code but with different missing argument to check

File: lib/dialects/postgres/query-generator.js [hunk 12]
```diff
@@ -564,24 +566,32 @@ const QueryGenerator = {
...
         if (Utils._.has(curParam, 'name')) { paramDef.push(curParam.name); }
         paramDef.push(curParam.type);
       } else {
-        throw new Error('createFunction called with a parameter with no type');
+        throw new Error('function or trigger used with a parameter without any type');
       }
-      return paramDef.join(' ');
+
+      var joined = paramDef.join(' ');
+      if(joined) paramList.push(joined);
```
Comment (sushantdhiman) [inline lib/dialects/postgres/query-generator.js]: missed space with `if`

File: test/integration/dialects/postgres/query-interface.test.js [hunk 13]
```diff
@@ -15,6 +17,163 @@ if (dialect.match(/^postgres/)) {
...
+      beforeEach( function () {
+        var self = this;
+        //make sure we don't have a pre-existing function called create_job
+        return self.queryInterface.dropFunction('create_job',[{type:'varchar',name:'test'}])
+        //suppress errors here. if create_job doesn't exist thats ok.
+        .catch( function (err) {  });
+      });
+
+      afterEach( function () {
+        var self = this;
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: Dont need `self`, you can just use `this`

File: test/integration/dialects/postgres/query-interface.test.js [hunk 14]
```diff
@@ -15,6 +17,163 @@ if (dialect.match(/^postgres/)) {
...
+      });
+
+      it('creates a stored procedure', function () {
+        var self = this;
+        var body = 'return test;';
+        var options = {};
+
+        //make our call to create a function
+        return new Promise.method( function () {
+          return self.queryInterface.createFunction('create_job', [{type:'varchar',name:'test'}], 'varchar', 'plpgsql', body, options);
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: Why we need to wrap it in `Promise.method`, `createFunction` already returns a promise

File: test/integration/dialects/postgres/query-interface.test.js [hunk 15]
```diff
@@ -15,6 +17,163 @@ if (dialect.match(/^postgres/)) {
...
+        })()
+        //validate
+        .then( function () { 
+          return self.sequelize.query('select create_job(\'test\');', { type: self.sequelize.QueryTypes.SELECT })
+          .then( function (res) {
+            if (res[0].create_job !== 'test') throw new Error('expected output: test');
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: rather than throw use `expect(res[0].create_job).to.be.eql('test')`

File: test/integration/dialects/postgres/query-interface.test.js (hunk 14 above)
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: This goes for other cases too where you have wrapped `create / rename / drop Function()` with `Promise.method`, They all return Promise already, no need to wrap them

File: lib/dialects/postgres/query-generator.js [hunk 16]
```diff
@@ -538,9 +539,10 @@ const QueryGenerator = {
   },
 
   dropFunction(functionName, params) {
+  if (!functionName) throw new Error("requires functionName");
```
Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: indentation

Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: `'` strings except for SQL

File: lib/dialects/postgres/query-generator.js [hunk 17]
```diff
@@ -564,24 +566,32 @@ const QueryGenerator = {
     return this.quoteIdentifier(Utils.removeTicks(this.escape(val), "'"));
   },
 
+
+
```
Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: ?

File: test/integration/dialects/postgres/query-interface.test.js [hunk 18]
```diff
@@ -8,13 +8,162 @@ var chai = require('chai')
...
+
+    describe('renameFunction', function () {
+
+      it('can rename a function');
+      it('requires parameters array');
+      it('requires old function name');
+      it('requires new function name');
+      it('throws an error if called with no parameter type');
+
+    });
```
Comment (mickhansen) [inline test/integration/dialects/postgres/query-interface.test.js]: ?

Comment (turbofoxwave) [inline test/integration/dialects/postgres/query-interface.test.js]: These are test stubs for tests I felt we should have. My apologies, the last project I was in we put in stubs. I'll remove them.

File: test/integration/dialects/postgres/query-interface.test.js [hunk 19]
```diff
@@ -8,13 +8,151 @@ var chai = require('chai')
...
+      it('can drop a function', function () {
+        var self = this;
+         
+        //call dropFunction
+        return self.queryInterface.dropFunction('droptest',[{type:'varchar',name:'test'}])
+        //validate
+        .then( function () {
+          return self.sequelize.query('select droptest(\'test\');', { type: self.sequelize.QueryTypes.SELECT })
+          .then( function () {
+            throw new Error('function failed to drop');
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: What is this code?  If function is dropped then `.then()` should never be triggered. it should automatically go to `catch` because we are trying to access a non existent function.

Also rather than `catch` and regex check, you can use the `expect.to.throw` here too.

File: test/integration/dialects/postgres/query-interface.test.js [hunk 20]
```diff
@@ -8,13 +8,151 @@ var chai = require('chai')
...
 if (dialect.match(/^postgres/)) {
   describe('[POSTGRES Specific] QueryInterface', function () {
     beforeEach(function () {
       this.sequelize.options.quoteIdenifiers = true;
       this.queryInterface = this.sequelize.getQueryInterface();
     });
 
+    describe('createFunction', function () {
+
+      beforeEach( function () {
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: `beforeEach` and `afterEach` both are doing the cleanup work and duplicating it. In each test we are creating a function that we should remove before next test. You can just get rid of `beforeEach` , because `afterEach` is already doing the cleanup work

File: test/integration/dialects/postgres/query-interface.test.js (hunk 19 above)
Comment (turbofoxwave) [inline test/integration/dialects/postgres/query-interface.test.js]: If droptest is not dropped as expected then  .then()  would be called because the droptest function will still exist.  The test will then produce an error stating the function failed to drop.

Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: Here is my understandings so far

- `dropFunction` removes the `droptest`
- Try to call the function `droptest` (note it has been dropped in last statement), So we are calling a dropped function. This statement should be rejected
- Add `catch` handler  or expect `error` to be thrown

But you are expecting `.then()` in Step 3, then throw error yourself to catch it again why ? If `dropFunction` really fails to remove the function in Step 1, mocha should catch the failed promise.

You should do something like

```js
return self.queryInterface
.dropFunction('droptest',[{type:'varchar',name:'test'}])
.then(function () {
  return expect(self.sequelize.query('select droptest(\'test\');', {
    type: self.sequelize.QueryTypes.SELECT
  }))
  .to.eventually.rejectedWith(/function droptest.* does not exist/);
});
```

Comment (felixfbecker) [inline test/integration/dialects/postgres/query-interface.test.js]: Guys we can use arrow functions now 😄

File: test/integration/dialects/postgres/query-interface.test.js (hunk 20 above)
Comment (turbofoxwave) [inline test/integration/dialects/postgres/query-interface.test.js]: if the test suite fails or stops unexpectedly causing afterEach's cleanup to not be called and the test suite is then restarted the cleanup in beforeEach will be needed.  What do you suggest?

Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: Yes, `beforeEach` is definitely a good choice to prevent failures in such cases

File: test/integration/dialects/postgres/query-interface.test.js [hunk 21]
```diff
@@ -6,15 +6,153 @@ var chai = require('chai')
...
+      it('creates a stored procedure', function () {
+        var self = this;
+        var body = 'return test;';
+        var options = {};
+
+        //make our call to create a function
+        return self.queryInterface.createFunction('create_job', [{type:'varchar',name:'test'}], 'varchar', 'plpgsql', body, options)
+        //validate
+        .then( function () { 
+          return self.sequelize.query('select create_job(\'test\');', { type: self.sequelize.QueryTypes.SELECT })
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: Please chain the promises properly so it doesn't become a [promise hell](https://medium.com/@pyrolistical/how-to-get-out-of-promise-hell-8c20e0ab0513). Also get rid of `self` (where possible) by using either `bind` or `arrow functions` (preferred way). Same goes for other code section too

File: test/integration/dialects/postgres/query-interface.test.js [hunk 22]
```diff
@@ -6,15 +6,145 @@ var chai = require('chai')
...
+      beforeEach( function () {
+        //make sure we don't have a pre-existing function called create_job
+        //this is needed to cover the edge case of afterEach not getting called because of an unexpected issue or stopage with the 
+        //test suite causing a failure of afterEach's cleanup to be called.
+        return this.queryInterface.dropFunction('create_job',[{type:'varchar',name:'test'}])
+        //suppress errors here. if create_job doesn't exist thats ok.
+        .catch( err => {});
+      });
+
+      afterEach( function () {
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: As we discussed `beforeEach` is definitely required to cover edge cases, but we can remove `afterEach`

File: changelog.md [hunk 23]
```diff
@@ -1,4 +1,6 @@
 # Future
+- [FIXED] sql syntax issue due to missing space in Postgresql query interface function "dropFunction"
```
Comment (sushantdhiman) [inline changelog.md]: Lets keep it simple, Like

``[FIXED] `dropFunction` syntax errors (PostgresSQL)``

File: changelog.md [hunk 24]
```diff
@@ -1,4 +1,6 @@
 # Future
+- [FIXED] sql syntax issue due to missing space in Postgresql query interface function "dropFunction" 
+- [FIXED] Postgresql query interface function "createFunction" and "dropFunction" through addressing output issue with function expandFunctionParamList
```
Comment (sushantdhiman) [inline changelog.md]: Please remove this statement, one chagelog entry is sufficient

File: test/integration/dialects/postgres/query-interface.test.js (hunk 22 above)
Comment (turbofoxwave) [inline test/integration/dialects/postgres/query-interface.test.js]: Without the afterEach the tests will not clean up after themselves. Is this ok?

Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: Ok then we should add `after` rather than `afterEach`. With this the test will cleanup after each suite and same code will not run in each case.

**Comment by sushantdhiman** [comment]
LGTM, Need @mickhansen's review before merging

**Comment by turbofoxwave** [comment]
Hey Sushantdhiman, 
I know I might still have more work to do after mickhansen's review  but I wanted to take the opportunity to say thank you for your time. I appreciate your insight and direction. Its been very helpful and a great learning opportunity. 

Best Regards,
-Mike

**Comment by sushantdhiman** [comment]
Thanks Mike, I was really impressed by your steadfastness and patience. Thank you for this contribution

**Comment by felixfbecker** [comment]
Needs rebase

File: package.json [hunk 25]
```diff
@@ -113,7 +113,6 @@
     "test-integration-mysql": "cross-env DIALECT=mysql npm run test-integration",
     "test-integration-postgres": "cross-env DIALECT=postgres npm run test-integration",
     "test-integration-postgres-native": "cross-env DIALECT=postgres-native npm run test-integration",
-    "test-integration-sqlite": "cross-env DIALECT=sqlite npm run test-integration",
```
Comment (sushantdhiman) [inline package.json]: Please dont remove the `sqlite` integration :)

File: test/integration/dialects/postgres/query-interface.test.js [hunk 26]
```diff
@@ -6,15 +6,145 @@ var chai = require('chai')
...
+        var options = {};
+
+        //make sure we have a droptest function in place.
+        return this.queryInterface.createFunction('droptest', [{type:'varchar',name:'test'}], 'varchar', 'plpgsql', body, options)
+        //suppress errors.. this could fail if the function is already there.. thats ok. 
+        .catch( function (err) { });
+      });
+
+      it('can drop a function', function () {
+
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: Remove this extra line

File: test/integration/dialects/postgres/query-interface.test.js [hunk 27]
```diff
@@ -6,15 +6,145 @@ var chai = require('chai')
...
+        .then( () => {
+          // call the function we attempted to drop.. if it is still there then throw an error informing that the expected behavior is not met. 
+          return this.sequelize.query('select droptest(\'test\');', { type: this.sequelize.QueryTypes.SELECT });  
+        }))
+        //test that we did get the expected error indicating that droptest was properly removed.
+        .to.be.rejectedWith(/.*function droptest.* does not exist/);
+      });
+
+      it('produces an error when missing expected parameters', function () {
+
```
Comment (sushantdhiman) [inline test/integration/dialects/postgres/query-interface.test.js]: Remove this extra line too
